### ISSUE TITLE
Wishlist (Notify the player when the character sees an item player wants)

### ISSUE THREAD

**Issue by Brambor** [CleverRaven/Cataclysm-DDA#51284]
# Is your feature request related to a problem? Please describe.
Usecase: I need X for a crafting recipe, I put it on the wishlist, next time I run through town, I might stumble on it.
It is actually possible to check seen items every turn and give X higher priority, this is an attempt for putting the burden of this optimization from the player on the engine and by doing so making it from impossible to pull off & have fun to a fun feature.

# Describe the solution you'd like
**The wishlist**
List items you want, when you see them, a prompt appears. "You see X from your wishlist!" or just "You see an item from your wishlist!". You then find the item through the view menu.

implementation of "see": item appears on a tile you see this turn but haven't seen the previous turn.
This includes items discovered by:
 - moving in a direction (usually) reveals new tiles
 - sun going up reveals new tiles
 - light sources (turn on a flashlight, zapper zombie in the night, fire) reveal new tiles

but the current algorithm wouldn't include
 - zombie dies and drops the item
 - something breaks / has been harvested / ... and drops the item

I'm asking for advice, mainly whether it is computably feasible.

Furthermore, a "no wishlist" zone should be added. Wishlist will simply not consider items in these zones.

# Describe alternatives you've considered
Put a widget to the sidebar showing the number of whitelisted items visible - unnecessary and while moving, the player would be incentivised to look at it constantly.

# Additional context
None

**Comment by Brambor** [comment]
@eltank on Discord:
> I'd say the hardest part is the UI for it. I would put the trigger in `game::place_player` which also triggers auto-pickup and auto-forage. To make it more efficient I'd put the desired item types in a `std::hash_set<const itype *>` (initialized after the user chooses an item filter) and then for all visible items I'd check if `item.type` is in that set

**Comment by chrispikula** [comment]
Hmmm.  What about tying that idea in further, and putting it into the auto pickup menu?  Have a 'alert' beside 'include' and 'exclude'?
This makes it enable/disable able, as well as provides a keyword test search interface, ie, the autopickup search test interface.

Furthermore, the idea would be that if you enable/disable autopickup, you could also enable/disable see & alert.  Maybe a tiny pop-up menu.  I'd want to see a possible cooldown on alerts on it as well, similar to the safe-mode alerts, given that I don't want to be informed every second that there's a widget that I'm looking for at my home base.  Maybe make it so that you need to travel a certain distance xy radially before it re-enables?

**Comment by Brambor** [comment]
> Hmmm. What about tying that idea in further, and putting it into the auto pickup menu? Have a 'alert' beside 'include' and 'exclude'?
> This makes it enable/disable able, as well as provides a keyword test search interface, ie, the autopickup search test interface.

I'm not really considering where to put the interface. The pickup menu seems like a good option though.

> Furthermore, the idea would be that if you enable/disable autopickup, you could also enable/disable see & alert. Maybe a tiny pop-up menu. I'd want to see a possible cooldown on alerts on it as well, similar to the safe-mode alerts, given that I don't want to be informed every second that there's a widget that I'm looking for at my home base. Maybe make it so that you need to travel a certain distance xy radially before it re-enables?

Making a "no wishlist zone" over your whole base is better than having any alerts in your base at all. I thought about the similarity to safe-mode, but I really don't think we should mix those. Realistically, you need 40 rocks, so you set an alert for rock. Then you want to be informed of every rock. If you want only one thing, one watch, then you wish for it until you find it. When you find it you disable the wish. No need for automatic disabling and reenabling then.

I will edit the post to include the zone thing.

**Comment by leemuar** [comment]
The game already has priority list in V menu. If you add some words to it - items will be listed in the top of the list of V menu

**Comment by Brambor** [comment]
> The game already has priority list in V menu. If you add some words to it - items will be listed in the top of the list of V menu. At least that's what I use to find items I interested in

Well, as I said in the second sentence:
> It is actually possible to check seen items every turn and give X higher priority, this is an attempt for putting the burden of this optimization from the player on the engine and by doing so making it from impossible to pull off & have fun to a fun feature.

When you run alongside a forest, you don't want to check every turn for... whatever... rocks. You just want to run until you can see a rock and pick it up. I hope we understand each other.

**Comment by leemuar** [comment]
Yeah, we do. I just wanted to make sure you know about the priority lists in current implementation

> hope we understand each other.

**Comment by matshou** [comment]
The lack of this feature was the first thing that bothered me as a new player.

**Comment by LeahLuong** [comment]
Agreed, would be really nice to have this feature. Would also be nice if OP corrected the spelling error in thread's title. I don't usually bug on such things but OP _**does**_ know the correct spelling, _**does**_ know how to edit their post, & the title is only 1 word. Maybe it's a pun on on how they're whistling into the wind (??) but still: spelling error.

**Comment by Brambor** [comment]
@LeahLuong Correction, OP is dyslectic and **_doesn't_** know the spelling (or even the spelling of the word dyslectic). But add-ons and word correct notify me of mistakes. Unless they don't catch it in time, because it's just one word,

Thanks for the correction!

**Comment by leemuar** [comment]
This is a good candidate for a new bionic module. Something like a scanner of sorts, it scans surroundings and alerts when desired item is found

**Comment by matshou** [comment]
Characters should be have access to this functionality without bionics modules because you don't need a bionic module to see.
However we could make a bionic module that allows you to scan for such items behind obstacles that block line of sight.

**Comment by leemuar** [comment]
It is open for debates how to implement it. Character can currently use V and priority list to search for items from "wishlist". Yes, it's more buttons to press (like it is something bad in CDDA...), but from my current perspective - perfectly "realistic" in CDDA design. That's why for automatic detection I suggest bionics.

Im not saying you're wrong, Im inviting to discuss it

> Characters should be have access to this functionality without bionics modules because you don't need a bionic module to see. However we could make a bionic module that allows you to scan for such items behind obstacles that block line of sight.

**Comment by Brambor** [comment]
> It is open for debates how to implement it. Character can currently use V and priority list to search for items from "wishlist". Yes, it's more buttons to press (like it is something bad in CDDA...), but from my current perspective - perfectly "realistic" in CDDA design. That's why for automatic detection I suggest bionics.
> 
> Im not saying you're wrong, Im inviting to discuss it

You are not first, so I will go into greater detail this time.

In all games, there are fun parts and routine parts. The game is fun as long as fun > routine. And is very fun when fun >> routine (much more than).

["Given the opportunity, players will optimize the fun out of a game."](https://www.designer-notes.com/?p=369) In CDDA you can choose to play the optimal way, which is to check the `V` menu _**every single movement**_, to spot that one rock you are looking for. I go for a less optimal way, roughly checking once every screen pass (like 50 steps). It is my sweet spot between wasting Characters time and my time.

But looking at it from a game design perspective, you don't need to check every N steps, you need to check only once per game. Just tell the game "notify me, when the Character can see a rock." That way you (almost) completely got rid of the routine part of finding the rock and you can focus on the fun part, like how will you carry this 20th rock without breaking your back.

It isn't some amazing new thing the Character could have not done before because of physical limitations. That's why I wouldn't put this feature behind the CBM wall which would prevent its availability for most of the time spent in the game. And most of this feature's useful time. Once you have CBMs, you don't need to search like that too much.

**Comment by matshou** [comment]
> Yes, it's more buttons to press (like it is something bad in CDDA...)

Not sure I understand this argument. Are you saying that adding more buttons to press without any real benefit is a good thing? Because if so then I would respectfully disagree. For me the appeal of the game comes from realism and the freedom to do things my way. The massive amount of button pressing is unfortunately the best design that we came up with so far to give players freedom inside the sandbox. We should always look for ways to reduce the amount of mechanical input required from player without sacrificing the level of design complexity.

**Comment by leemuar** [comment]
@Brambor @matshou I agree with you and I understand your point. I want this feature too.

The thing is "fun" is relative. We have a diverse community and different people enjoy different things. Some people like micromanagement. Some don't.

My question is based on my experience on the game's discord discussion. I once whined about "too many buttons to do X". I was told "pressing lots of buttons in CCDA is fun". So Im kinda in constant doubt about what is right or wrong for the game in terms of it's design, because people see it differently.

Again, Im not against the "Hey, I see that item over there!" implementation, actually Im in for it. I just invite to discuss it.

**Comment by leemuar** [comment]
I suggest to set the issue title to something like "Notify the player when character sees the item player wants"

**Comment by Night-Pryanik** [comment]
Closing as stale, since stalebot can't do this by itself.
